pkg/postage/batchstore/reserve.go

```diff
@@ -0,0 +1,230 @@
+// in the relevant POs, belonging to the batch and (unless they are otherwise pinned)
+// allow  them  to be gargage collected.
+//
+// the rules of the reserve
+// - if batch a is unreserved and val(b) <  val(a) then b is unreserved on any po
+// - if a batch is unreserved on po  p, then  it is unreserved also on any p'<p
+// - batch size based on fully filled the reserve should not  exceed Capacity
+// - batch reserve is maximally utilised, i.e, cannot be extended and have 1-3 remain true
+package batchstore
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/ethersphere/bee/pkg/postage"
+	"github.com/ethersphere/bee/pkg/swarm"
+)
+
+// ErrBatchNotFound is returned when the postage batch is not found or expired
+var ErrBatchNotFound = errors.New("postage batch not found or expired")
+
+// DefaultDepth is the initial depth for the reserve
+const DefaultDepth = 5
+
+// Capacity = number of chunks in reserve
+var Capacity = exp2(23)
+
+var big1 = big.NewInt(1)
+
+// reserveState records the state and is persisted in the state store
+type reserveState struct {
+	Depth    uint8    `json:"depth"`    // Radius of responsibility
+	Capacity int64    `json:"capacity"` // size of the reserve -  number of chunks
+	Edge     *big.Int `json:"edge"`     // lower value limit for edge layer = the further half of chunks
+	Core     *big.Int `json:"core"`     // lower value limit for core layer = the closer half of chunks
+}
+
+// unreserve is called when the batchstore decides not to reserve a batch on a PO
+// i.e.  chunk of  the batch in bins [0 upto PO] (closed  interval) are unreserved
+func (s *store) unreserve(batchID []byte, radius uint8) error {
+	return s.unreserveFunc(batchID, radius)
+}
+
+func (s *store) purgeExpired() error {
+	var toDelete [][]byte
+	err := s.store.Iterate(valueKeyPrefix, func(key, _ []byte) (bool, error) {
+		b, err := s.Get(valueKeyToID(key))
+		if err != nil {
+			return true, err
+		}
+		if b.Value.Cmp(s.rs.Core) >= 0 {
+			return true, nil
+		}
+		err = s.unreserve(b.ID, swarm.MaxPO)
+		if err != nil {
+			return true, err
+		}
+		// if batch has no value then delete it
+		if b.Value.Cmp(s.cs.Total) <= 0 {
+			toDelete = append(toDelete, b.ID)
+		} else {
+			b.Radius = swarm.MaxPO
+			if err := s.store.Put(batchKey(b.ID), b); err != nil {
+				return true, err
+			}
+		}
+
+		return false, nil
+	})
+	if err != nil {
+		return err
+	}
+	return s.delete(toDelete...)
+}
+
+type change struct {
+	old, new int
+}
+
+func newChange(oldv, newv, lower, higher *big.Int) *change {
+	return &change{
+		old: cmp(oldv, lower, higher),
+		new: cmp(newv, lower, higher),
+	}
+}
+
+func cmp(x, lower, higher *big.Int) int {
+	if x.Cmp(lower) < 0 || x.Cmp(big.NewInt(0)) == 0 {
+		return -1
+	}
+	if x.Cmp(higher) < 0 {
+		return 0
+	}
+	return 1
+}
+
+func (ch *change) changed() bool {
+	return ch.old != ch.new
+}
+
+func (ch *change) increased() bool {
+	return ch.old < ch.new
+}
+
+func (ch *change) double() bool {
```

Comment:
why does double mean that the sum is zero?

Reply:
the change is 'double' if it  goes from not in reserve to edge-reserved or the other way round